File: tests/fixtures/project_py/src/utils/validation.py
from __future__ import annotations

import re

from src.utils.errors import ValidationError


_EMAIL_REGEX = re.compile(r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$")
_MIN_PASSWORD_LENGTH = 8


def validate_email(email: str) -> str:
    """Validate an email address format. Returns the normalized email or raises ValidationError."""
    email = email.strip().lower()
    if not _EMAIL_REGEX.match(email):
        raise ValidationError(field="email", reason="Invalid email format")
    return email


def validate_password(password: str) -> str:
    """Validate password strength. Returns the password or raises ValidationError."""
    if len(password) < _MIN_PASSWORD_LENGTH:
        raise ValidationError(
            field="password",
            reason=f"Password must be at least {_MIN_PASSWORD_LENGTH} characters",
        )
    if not re.search(r"[A-Z]", password):
        raise ValidationError(field="password", reason="Password must contain an uppercase letter")
    if not re.search(r"[0-9]", password):
        raise ValidationError(field="password", reason="Password must contain a digit")
    return password


def sanitize_input(value: str) -> str:
    """Strip dangerous characters from user input."""
    sanitized = re.sub(r"[<>&\"']", "", value)
    return sanitized.strip()

File: tests/fixtures/project_go/utils/validation.go
package utils

import (
	"regexp"
	"strings"
)

var emailRegex = regexp.MustCompile(`^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$`)

// ValidateEmail checks whether the given string is a valid email address.
func ValidateEmail(email string) bool {
	return emailRegex.MatchString(email)
}

// ValidatePassword checks whether the password meets minimum requirements.
// Returns true if the password is at least 8 characters long.
func ValidatePassword(password string) bool {
	return len(password) >= 8
}

// SanitizeInput trims whitespace and removes null bytes from the input string.
func SanitizeInput(input string) string {
	sanitized := strings.TrimSpace(input)
	sanitized = strings.ReplaceAll(sanitized, "\x00", "")
	return sanitized
}
